### PR TITLE
Add wburls to keep track of urls in the SDK

### PR DESCRIPTION
**Pull request by raubitsj**
Description
-----------
Add some process around putting urls in the SDK.   This way we can add a checker (future) which makes sure that the links always resolve.

Testing
-------
manually tried invalid things to make sure mypy complains.

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/lib/wburls.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    def _get_urls(self) -> Dict["URLS", str]:
+        return dict(
+            cli_launch="http://wandb.me/launch",
+            doc_run="https://wandb.me/run",
+            doc_require="http://wandb.me/library-require",
+            doc_start_err="https://docs.wandb.ai/library/init#init-start-error",
+            upgrade_local="https://wandb.me/local-upgrade",
+        )
+
+    def get(self, s: "URLS") -> str:
```
Comment (kptkin) [inline wandb/sdk/lib/wburls.py:36]: nit: it would be nice to use `__getitem__` for convince, what do you think?

File: wandb/sdk/lib/wburls.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+from typing import Dict, Optional
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from ._wburls_generated import URLS
+
+
+class WBUrls:
```
Comment (kptkin) [inline wandb/sdk/lib/wburls.py:21]: nit: feel weird that the url is not all capitalized

File: tools/generate-tool.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+GENERATE_SUFFIX = "_generate.py"
+GENERATED_SUFFIX = "_generated.py"
+
+
+def get_paths():
+    paths = []
+    if not args.files:
+        exclude_dirs = {"vendor", "__pycache__"}
+        root_dir = pathlib.Path(__file__).resolve().parent.parent / "wandb"
```
Comment (kptkin) [inline tools/generate-tool.py:44]: seems fragile way of specifying the root dir...

File: tools/generate-tool.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        print(f"INFO: Generating {output_path}...")
+        generate_file(p, output_path)
+
+
+def format_file(filename):
+    status, output = subprocess.getstatusoutput(f"black {filename}")
+    assert status == 0, f"Error: {output}"
+
+
+def format_files(paths):
```
Comment (kptkin) [inline tools/generate-tool.py:76]: seems simple enough to be in a single function?

File: tools/generate-tool.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+def generate_file(generate_path, output_path):
+    status, output = subprocess.getstatusoutput(f"python {generate_path}")
+    assert status == 0, f"Error: {output}"
+    with open(output_path, "w") as f:
+        f.write("# DO NOT EDIT -- GENERATED BY: `generate-tool.py --generate`")
+        f.write(output)
+
+
+def generate_files(paths):
```
Comment (kptkin) [inline tools/generate-tool.py:64]: should we merge this 2?

**Comment by kptkin** [comment]
could you add a small documentation for the process of adding a new wandb url?

**Comment by kptkin** [comment]
like the idea of checking link resolve... 😄

File: tools/generate-tool.py (hunk 4 above)
Comment (raubitsj) [reply tools/generate-tool.py:76]: it is reused in check.. thats why it is split out.

File: wandb/sdk/lib/wburls.py (hunk 2 above)
Comment (raubitsj) [reply wandb/sdk/lib/wburls.py:21]: yeah,  but then it is all caps and flake and other tools complain that you shouldnt make your thing all caps.  the problem with using a class name with two initialisms :)

File: wandb/sdk/lib/wburls.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk/lib/wburls.py:36]: i think this is fine.. more explicit the better also it forces you not to specify a default, i want it to assert on not found

File: tools/generate-tool.py (hunk 5 above)
Comment (raubitsj) [reply tools/generate-tool.py:64]: Nope.. it is reused in check_files

**Comment by raubitsj** [comment]
> could you add a small documentation for the process of adding a new wandb url? (maybe in contribute?)

done

File: wandb/sdk/lib/wburls.py (hunk 2 above)
Comment (kptkin) [reply wandb/sdk/lib/wburls.py:21]: what about `WBURLs`? no strong feelings... :)

Comment (raubitsj) [reply wandb/sdk/lib/wburls.py]: SURE!  done

File: wandb/sdk/lib/_wburls_generate.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+URLS = Literal[
+    $literal_list
+]
+"""
+
+
+def generate() -> None:
+    urls = wburls._get_urls()
+    literal_list = ", ".join([f'"{key}"' for key in urls])
+    print(template.replace("$literal_list", literal_list))
```
Comment (dmitryduev) [inline wandb/sdk/lib/_wburls_generate.py:21]: this is really nice!

**Review by dmitryduev — APPROVED** [review]

**Pull request merged**